pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,835 @@
+    """A quantum tape recorder, that records, validates, executes,
+    and differentiates variational quantum programs.
+
+    .. note::
+
+        As the quantum tape is a *beta* feature, the standard PennyLane
+        measurement functions cannot be used. You will need to instead
+        import modified measurement functions within the quantum tape:
+
+        >>> from pennylane.beta.queuing import expval, var, sample, probs
+
+    **Example**
+
+    .. code-block:: python
+
+        from pennylane.beta.tapes import QuantumTape
+        from pennylane.beta.queuing import expval, var, sample, probs
+
+        with QuantumTape() as tape:
+            qml.RX(0.432, wires=0)
+            qml.RY(0.543, wires=0)
+            qml.CNOT(wires=[0, 'a'])
+            qml.RX(0.133, wires='a')
+            expval(qml.PauliZ(wires=[0]))
+
+    Once constructed, information about the quantum circuit can be queried:
+
+    >>> tape.operations
+    [RX(0.432, wires=[0]), RY(0.543, wires=[0]), CNOT(wires=[0, 'a']), RX(0.133, wires=['a'])]
+    >>> tape.observables
+    [expval(PauliZ(wires=[0]))]
+    >>> tape.get_parameters()
+    [0.432, 0.543, 0.133]
+    >>> tape.wires
+    <Wires = [0, 'a']>
+    >>> tape.num_params
+    3
+
+    The :class:`~.beta.tapes.NewCircuitGraph` can also be accessed:
+
+    >>> tape.graph
+    <pennylane.beta.tapes.circuit_graph.NewCircuitGraph object at 0x7fcc0433a690>
+
+    Once constructed, the quantum tape can be executed directly on a supported
+    device:
+
+    >>> dev = qml.device("default.qubit", wires=[0, 'a'])
+
+    Execution can take place either using the in-place constructed parameters,
+
+    >>> tape.execute(dev)
+    [0.77750694]
+
+    or by providing parameters at run time:
+
+    >>> tape.execute(dev, params=[0.1, 0.1, 0.1])
+    [0.99003329]
+
+    The Jacobian can also be computed using finite difference:
+
+    >>> tape.jacobian(dev)
+    [[-0.35846484 -0.46923704  0.        ]]
+    >>> tape.jacobian(dev, params=[0.1, 0.1, 0.1])
+    [[-0.09933471 -0.09933471  0.        ]]
+
+    Finally, the trainable parameters can be explicitly set, and the values of
+    the parameters modified in-place:
+
+    >>> tape.trainable_params = {0} # set only the first parameter as free
+    >>> tape.set_parameters(0.56)
+    >>> tape.get_parameters()
+    [0.56]
+    >>> tape.get_parameters(trainable_only=False)
+    [0.56, 0.543, 0.133]
+
+    Trainable parameters are taken into account when calculating the Jacobian,
+    avoiding unnecessary calculations:
+
+    >>> tape.jacobian(dev)
+    [[-0.45478169]]
+    """
+
+    def __init__(self, name=None):
+        super().__init__()
+        self.name = name
+
+        self._prep = []
+        """list[~.Operation]: Tape state preparations."""
+
+        self._ops = []
+        """list[~.Operation]: quantum operations recorded by the tape."""
+
+        self._measurements = []
+        """list[~.MeasurementProcess]: measurement processes recorded by the tape."""
```

Comment:
Just a thought: With this design we will never be able to support intermediate measurements, no? Which is fine because they can always be moved to the end. But what happens in the case of intermediate measurements that condition future gates on measurment results? Can those also be moved to the end? I'm thinking of a measurement-based QC design...

Reply:
Yep I had this same thought as well. There are so many assumptions everywhere regarding measurements being at the end of the circuit.... I didn't know whether to just stick with it for now (less work) or change it

Reply:
Perhaps we could add tapes that support intermediate measurements at some points... allows us to keep both behaviour for a while, and avoids massive refactors